### PR TITLE
Support v2 or v3 for SDS

### PR DESCRIPTION
**Pull request by howardjohn**
For https://github.com/istio/istio/issues/19885

This will be a bit easier with https://github.com/istio/istio/pull/24288 so I wil wait for that PR to land

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/adapter.go [hunk 1]
```diff
@@ -60,7 +59,21 @@ func (d DiscoveryStreamV2Adapter) Recv() (*discovery.DiscoveryRequest, error) {
 		TypeUrl:       v2Req.TypeUrl,
 		ResponseNonce: v2Req.ResponseNonce,
 		ErrorDetail:   v2Req.ErrorDetail,
-	}, err
+	}
+}
+
+// Convert from v2 to v3
```
Comment (hzxuzhonghu) [inline pilot/pkg/proxy/envoy/v2/adapter.go:65]: v3->v2?

File: security/pkg/nodeagent/sds/adapter.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import (
+	"context"
+
+	xdsapiv2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	discoveryv2 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	v2 "istio.io/istio/pilot/pkg/proxy/envoy/v2"
```
Comment (hzxuzhonghu) [inline security/pkg/nodeagent/sds/adapter.go:11]: maybe should rename to envoy/v2 later

File: security/pkg/nodeagent/sds/adapter.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	s *sdsservice
+}
+
+func (d *sdsServiceAdapter) FetchSecrets(ctx context.Context, request *xdsapiv2.DiscoveryRequest) (*xdsapiv2.DiscoveryResponse, error) {
+	v3Resp, err := d.s.FetchSecrets(ctx, v2.UpgradeV2Request(request))
+	return v2.DowngradeV3Response(v3Resp), err
+}
+
+// We implement the v2 ADS API
+var _ discoveryv2.SecretDiscoveryServiceServer = &sdsServiceAdapter{}
```
Comment (JimmyCYJ) [inline security/pkg/nodeagent/sds/adapter.go:40]: What is the purpose of this step?

Comment (howardjohn) [reply security/pkg/nodeagent/sds/adapter.go:40]: support backwards compatibly with old pilot

**Pull request merged**